### ISSUE TITLE
Unrelated events triggered in Windows

### ISSUE THREAD

**Issue by gustavohenke** [paulmillr/chokidar#231]
I found this problem originally by using gulp-watch, but it seems like its actually related to chokidar.
So, whenever I change/add some file, other unrelated events are triggered as well. I'm running Windows 8.1, and using some server that runs Ubuntu 13.10 I don't see this problem.

My test file:

```js
require('chokidar').watch('*').on('all', function(event, path) {
  console.log(event, path);
});
```

My dependencies:
```
@ C:\Users\Gustavo\Projetos\tst
└── chokidar@1.0.0-rc3
```

I recorded a video of the problem:
https://docs.google.com/file/d/0B5uvgO7NPTE9LS1BY0ktUFczQVk/edit?usp=drivesdk

Related issue: floatdrop/gulp-watch#133

**Comment by es128** [comment]
So this issue is specific to new files you've created getting events emitted at the directory's `fs.watch` instance whenever you add another file. Ultimately it's a bug in `fs.watch`, but we might be able to correct for it in chokidar. I am guessing this is related to the fix for a different issue involving the vim editor, so we can try filtering that fix to not apply to windows. 

Please try installing my proposed fix with `npm i paulmillr/chokidar#issue-231`.

If you're still able to reproduce the issue, please change your test script to the following to help with diagnosing what's actually happening:

```js
require('chokidar').watch('*').on('all', function(event, path) {
  console.log(event, path);
}).on('raw', console.log);
```

**Comment by gustavohenke** [comment]
Still not worked, even using your test script:

```
@ C:\Users\Gustavo\Projetos\tst
└── chokidar@1.0.0-rc3 (git://github.com/paulmillr/chokidar#5047c37033ca07ca48ce195c4e65560c45aff2c7)
```

Video: https://docs.google.com/file/d/0B5uvgO7NPTE9em9FUGhxZ0hxRDQ/edit?usp=drivesdk

Note that even changing existing files or removing them leads to these events.

**Comment by es128** [comment]
As you can see from the raw event logs, the system is actually emitting change events for all those files every time you write to a new one. I don't see any way to filter out the illegitimate ones.

Try with an actual text editor. And try giving the files more legit names - the similarity in the filenames may be contributing to the issue.

> Note that even changing existing files or removing them leads to these events.

You haven't shown that. The two files that were already in the directory when you started the watcher are not getting these extra events. Only the ones you're piping into from the shell after the watcher started.

**Comment by gustavohenke** [comment]
The strange thing is that while using gulp-watch v3.0.0 (which uses gaze instead - but is very slower), I never had these problems.

Look, this is the log of my app using gulp-watch v4.1.0:

```
[13:14:32] home\home.html was changed
[13:14:32] Starting 'templates'...
[13:14:33] home\evento-date.filter.js was unlinked
[13:14:33] Starting 'js'...
[13:14:33] home\evento.directive.js was unlinked
[13:14:33] home\followup-route.service.js was unlinked
[13:14:33] home\followup.controller.js was unlinked
[13:14:33] home\home.controller.js was unlinked
[13:14:33] home\routes.config.js was unlinked
[13:14:34] home\evento-date.filter.js was added
[13:14:34] home\evento.directive.js was added
[13:14:34] home\followup-route.service.js was added
[13:14:34] home\routes.config.js was added
[13:14:34] home\evento-date.filter.js was unlinked
[13:14:34] home\evento.directive.js was unlinked
[13:14:34] home\followup-route.service.js was unlinked
[13:14:34] home\followup.controller.js was unlinked
[13:14:34] home\home.controller.js was unlinked
[13:14:34] home\routes.config.js was unlinked
[13:14:34] home\home.controller.js was added
[13:14:34] home\followup.controller.js was added
[13:14:34] home\evento-date.filter.js was changed
[13:14:34] home\evento.directive.js was changed
[13:14:34] home\followup-route.service.js was changed
[13:14:34] home\followup.controller.js was changed
[13:14:34] home\home.controller.js was changed
[13:14:34] home\routes.config.js was changed
[13:14:34] home\evento-date.filter.js was changed
[13:14:34] home\evento.directive.js was changed
[13:14:34] home\followup-route.service.js was changed
[13:14:34] home\followup.controller.js was changed
[13:14:34] home\home.controller.js was changed
[13:14:34] home\routes.config.js was changed
[13:14:34] home\evento-date.filter.js was added
[13:14:34] home\evento.directive.js was added
[13:14:34] home\followup-route.service.js was added
[13:14:34] home\followup.controller.js was added
[13:14:34] home\home.controller.js was added
[13:14:34] home\routes.config.js was added
[13:14:34] home\evento.directive.js was changed
[13:14:34] home\evento-date.filter.js was changed
[13:14:34] home\followup-route.service.js was changed
[13:14:34] home\followup.controller.js was changed
[13:14:34] home\home.controller.js was changed
[13:14:34] home\routes.config.js was changed
[13:14:34] Finished 'js' after 1.56 s
[13:14:35] config\config.controller.js was unlinked
[13:14:35] Starting 'js'...
[13:14:35] config\routes.config.js was unlinked
[13:14:35] evento\acao.directive.js was unlinked
[13:14:35] evento\acoes-evento.controller.js was unlinked
[13:14:35] evento\cliente-evento.controller.js was unlinked
[13:14:35] evento\evento.controller.js was unlinked
[13:14:35] evento\routes.config.js was unlinked
[13:14:35] core\appGlobals.provider.js was unlinked
[13:14:35] core\state-loader.service.js was unlinked
[13:14:35] home\evento.directive.js was unlinked
[13:14:35] home\evento-date.filter.js was unlinked
[13:14:35] home\followup-route.service.js was unlinked
[13:14:35] home\followup.controller.js was unlinked
[13:14:35] home\home.controller.js was unlinked
[13:14:35] home\routes.config.js was unlinked
[13:14:35] login\login-error.controller.js was unlinked
[13:14:35] login\login.controller.js was unlinked
[13:14:35] login\routes.config.js was unlinked
[13:14:35] core\appGlobals.provider.js was unlinked
[13:14:35] core\state-loader.service.js was unlinked
[13:14:35] login\login-error.controller.js was unlinked
[13:14:35] login\login.controller.js was unlinked
[13:14:35] login\routes.config.js was unlinked
[13:14:35] evento\acao.directive.js was unlinked
[13:14:35] evento\acoes-evento.controller.js was unlinked
[13:14:35] evento\cliente-evento.controller.js was unlinked
[13:14:35] evento\evento.controller.js was unlinked
[13:14:35] evento\routes.config.js was unlinked
[13:14:35] config\config.controller.js was unlinked
[13:14:35] config\routes.config.js was unlinked
[13:14:35] home\evento-date.filter.js was unlinked
[13:14:35] home\evento.directive.js was unlinked
[13:14:35] home\followup-route.service.js was unlinked
[13:14:35] home\followup.controller.js was unlinked
[13:14:35] home\home.controller.js was unlinked
[13:14:35] home\routes.config.js was unlinked
[13:14:35] config\config.controller.js was added
[13:14:35] config\routes.config.js was added
[13:14:35] core\appGlobals.provider.js was added
[13:14:35] evento\acao.directive.js was added
[13:14:35] evento\acoes-evento.controller.js was added
[13:14:35] evento\evento.controller.js was added
[13:14:35] evento\routes.config.js was added
[13:14:35] evento\cliente-evento.controller.js was added
[13:14:35] login\login-error.controller.js was added
[13:14:35] login\routes.config.js was added
[13:14:35] home\evento-date.filter.js was added
[13:14:35] home\evento.directive.js was added
[13:14:35] home\followup.controller.js was added
[13:14:35] home\home.controller.js was added
[13:14:35] home\routes.config.js was added
[13:14:35] login\login.controller.js was added
[13:14:35] home\followup-route.service.js was added
[13:14:35] core\appGlobals.provider.js was added
[13:14:35] core\state-loader.service.js was added
[13:14:35] config\config.controller.js was added
[13:14:35] config\routes.config.js was added
[13:14:35] home\evento-date.filter.js was added
[13:14:35] home\evento.directive.js was added
[13:14:35] home\followup-route.service.js was added
[13:14:35] home\followup.controller.js was added
[13:14:35] home\home.controller.js was added
[13:14:35] home\routes.config.js was added
[13:14:35] login\login-error.controller.js was added
[13:14:35] login\routes.config.js was added
[13:14:35] evento\acao.directive.js was added
[13:14:35] evento\acoes-evento.controller.js was added
[13:14:35] evento\cliente-evento.controller.js was added
[13:14:35] evento\evento.controller.js was added
[13:14:35] evento\routes.config.js was added
[13:14:35] login\login.controller.js was added
[13:14:36] core\state-loader.service.js was added
[13:14:36] Finished 'js' after 1.32 s
[13:14:36] Finished 'templates' after 4.47 s
```

_All_ I did was change the `home.html` file.

You can see more videos for such problems in the gulp-watch issue I reported: floatdrop/gulp-watch#133

**Comment by es128** [comment]
What editor are you using?

I want to keep getting to the bottom of this, but as a short-term solution you could also try switching to polling mode (`usePolling: true`)

**Comment by es128** [comment]
Also, what's happening in your `templates` task that might be touching things? Seems like the watcher thinks the entire directory gets unlinked and then comes back.

**Comment by es128** [comment]
Try the branch again

**Comment by gustavohenke** [comment]
This is the `templates` task (all watched sources are inside `src/`, output files are inside `www/`, so no conflicts here):

```js
gulp.task( "templates", function () {
    var files = gulp.src([
        "src/**/*.html",
        "!src/index.html",
        "!src/**/fixtures/*.html"
    ], {
        base: "."
    });
    return files.pipe( $.plumber() )
                .pipe( $.sourcemaps.init() )
                .pipe( $.ngTemplate( "templates.js", {
                    module: "app",
                    base: "src"
                }))
                .pipe( $.sourcemaps.write( ".", {
                    sourceRoot: "."
                }))
                .pipe( gulp.dest( "www" ) )
                .pipe( $.connect.reload() );
});
```

Using polling, the CPU usage increased to around 80%, but it "worked" (the first time I modified a file, same problem, but then onwards, fine).

I'll try with that branch again now.

**Comment by es128** [comment]
I wonder if there's a chance you have a virus scanner or some other sort of system-level utility which is having an impact.

**Comment by gustavohenke** [comment]
I did have, disabled and no success, even with the last commit you made.
I also tested in a Windows 7 Ultimate that's available in my company's network. Same problem.

**Comment by gustavohenke** [comment]
BTW, I tried various editors - Atom, IntelliJ IDEA, Sublime, CLI and even the ~~good and~~ old notepad.exe.

**Comment by es128** [comment]
There's seems to be something unique going on in your environment. Would be interested to hear from anyone else that can reproduce this.

**Comment by gustavohenke** [comment]
I created an repository which reproduces this issue:
https://github.com/gustavohenke/watch-events-problem

It's an slice of another project I have (which I can share with you because I'm the owner, not my company). The events keep happening, just in minor scale.

**Comment by es128** [comment]
I was able to kind of reproduce it using your repo in a windows VM, but not when just using chokidar to watch the same directory and do the same thing.

![image](https://cloud.githubusercontent.com/assets/1740641/6193126/4236a738-b386-11e4-94c5-91f7566661df.png)

Please try to isolate the problem so it's reproducible using chokidar only. At the moment my impression is that the bug is somewhere outside of chokidar.

**Comment by gustavohenke** [comment]
@floatdrop, any chance that's related with gulp-watch?

**Comment by floatdrop** [comment]
@gustavohenke probably not. `gulp-watch` is only a simple wrapper around watching library. Only funny thing I get from `watch-events-problem` repository - is that touching js files causing duplicate event:

```js
require('chokidar').watch([
    "src/**/*.module.js",
    "src/**/*.js"
  ]).on('all', function(event, path) {
  console.log(event, path);
});
```

And after start I get this:

```
add src/app.module.js
add src/app.module.js
add src/core/core.module.js
add src/home/home.module.js
add src/core/core.module.js
add src/core/init-routes.config.js
add src/core/theme.config.js
add src/home/home.controller.js
add src/home/home.module.js
add src/home/routes.config.js
add src/shared/i18n/i18n.module.js
add src/shared/i18n/i18n-loader.service.js
add src/shared/i18n/i18n.module.js
```

Seems like chokidar is not deduping files in some sort. Could it be, that arrays of globs is not supported by chokidar @es128?

**Comment by es128** [comment]
Well that first glob in your example is entirely redundant, but yeah those duped events are no good.

Actually in my test above I was just watching the directory, I didn't use the same array as in @gustavohenke's repo. In fact I see now that several watchers are being opened against the same base dir - so that's probably where the issue lies.

@floatdrop Do separate calls to gulp-watch result in separate chokidar FSWatcher instances, or do you use the `.add()` method? It is probably necessary to do the latter in order to get the correct behavior - or add your own de-duping logic. Not sure if you intend to support a use-case like the one being presented here.

**Comment by heikki** [comment]
Is there any sane reason why deduping would be necessary?

**Comment by es128** [comment]
@heikki Are you saying we should put it on users to not pass in redundant glob patterns as in @floatdrop's example a few comments up? At least within the same chokidar.FSWatcher instance I think I'd rather just handle it I can do so without too much complexity. 

Probably something along the lines of being smarter about using both glob filters together with one instance of the watcher against their common base dir.

**Comment by es128** [comment]
That is, unless I've already solved that particular issue as [this test](https://github.com/paulmillr/chokidar/blob/master/test.js#L476) would seem to indicate.

It's possible the whole problem here is separate watchers being opened to look at an overlapping set of files. I'm not sure yet, it's a tricky example.

**Comment by es128** [comment]
Well actually @floatdrop's example appears to indicate that particular problem is not actually solved and the existing test is insufficient.

**Comment by heikki** [comment]
@es128 Just wondering.. within the same instance deduping might make sense because people might expect that array of globs works the same way as in [gulp.src](https://github.com/wearefractal/vinyl-fs#srcglobs-opt). Maybe it would be good thing after all.

**Comment by gustavohenke** [comment]
I've used used that redundant glob because angular modules (*.module.js) need to be defined before they're used (other files).

**Comment by es128** [comment]
Is that something `gulp.dest` respects? I wouldn't count on it for getting file events in any particular order from chokidar.

**Comment by gustavohenke** [comment]
Yeah, when I concat them, the order will be respected.
Same applies for usemin kind of thing, that I'm also using.

**Comment by es128** [comment]
Ok, good to know. I'm still concerned about how you spawn several watchers, but as a first step I'll figure out how to fix the issue within a single watcher instance and then we'll see how things look from there.

**Comment by floatdrop** [comment]
> Do separate calls to gulp-watch result in separate chokidar FSWatcher instances
@es128 yes.

**Comment by gustavohenke** [comment]
@es128 Is there any way I could help fix this issue?
It's being annoying to use an ultra slow but effective gaze instead of a fast but _buggy-for-me_ chokidar :cry:

**Comment by es128** [comment]
PRs with failing tests and/or fixes welcome.

**Comment by gustavohenke** [comment]
So, I feel like I'm unable to further help if not via local testing. That's frustrating :disappointed: 

I recorded yet another video, this time it's a chokidar isolated issue:
https://docs.google.com/file/d/0B5uvgO7NPTE9cEdXV0E4cDFCTW8/edit?usp=drivesdk

The test files have been pushed again to [watch-events-problem](/gustavohenke/watch-events-problem) repository.

**Comment by es128** [comment]
https://github.com/paulmillr/chokidar/issues/231#issuecomment-74310387

@floatdrop was the outcome you described in this comment on Windows?

**Comment by floatdrop** [comment]
@es128 no, OS X 10.10.2

**Comment by es128** [comment]
I had trouble reproducing. Could you try with rc4? Maybe a different bug fix had an impact on this issue.

**Comment by floatdrop** [comment]
@es128 yes, this seems to be fixed on `chokidar@1.0.0-rc4`. Can not reproduce it either.

**Comment by es128** [comment]
@gustavohenke can you try with rc4 please?

**Comment by gustavohenke** [comment]
Same problems.

**Comment by es128** [comment]
I was able to reproduce this using your tst.js script. I tried to tweak the intersecting globs test to get it to fail so this case was covered, but didn't find a way to do so. Nonetheless, I believe I have fixed the issue with 1b99ad2. 

@gustavohenke please give it a try and provide feedback

**Comment by gustavohenke** [comment]
It seems to be working fine for now. Thank you very much @es128!

**Comment by robario** [comment]
I have the same problem. I have solved it by separating globs into a chunk as bellow.

```js
(Array.isArray(src) ? src : [src]).forEach(function (pattern) {
    require('gulp-watch')(pattern, {verbose: true});
});
```

**Comment by jasonkenneth** [comment]
Thanks for the great product! Is this fix published to npm, yet? I get this behavior when more than one inclusive glob is in the paths array.

random extra events:

        allAssets: [
            '*.json',
            '**/*.ico',
            '**/*.jpg',
            '**/*.png',
            '!tsd.json'
        ]

looks ok:

        allAssets: [
            '*.json',
            '!tsd.json'
        ]

**Comment by es128** [comment]
This fix hasn't been published yet. I'm trying to get any remaining non-esoteric issues addressed and get 1.0.0 published - after which ordinary semver can take over.  

But this rc phase has been very useful for uncovering some warts related to glob handling which is the most substantial new change compared with 0.12 and hadn't seen much real-world usage. I didn't want to unleash it widely too hastily when even small issues had a chance of impacting a lot of people.

**Comment by jasonkenneth** [comment]
Understood. Thanks, Elan. And thanks for the quick response.
